### PR TITLE
Add instrumentation to detect the route at the beginning of the spring request

### PR DESCRIPTION
**Pull request by tylerbenson**
Instead of waiting till the handler is called, otherwise if a response is returned by a filter then the proper name wouldn't be set and would fall back to the URL.

I tried to add comments for the parts that are interdependent.  Let me know if it's insufficient.

### CODE REVIEW

File: dd-java-agent/instrumentation/spring-webmvc-3.1/src/main/java/datadog/trace/instrumentation/springweb/HandlerMappingResourceNameFilter.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import org.springframework.web.servlet.HandlerExecutionChain;
+import org.springframework.web.servlet.HandlerMapping;
+
+public class HandlerMappingResourceNameFilter implements Filter {
```
Comment (devinsba) [inline dd-java-agent/instrumentation/spring-webmvc-3.1/src/main/java/datadog/trace/instrumentation/springweb/HandlerMappingResourceNameFilter.java:17]: I guess this doesn't have to implement Filter since it's not being used as one, but since we don't have access to the functional interfaces there doesn't really seem to be a better type in my mind

Comment (tylerbenson) [reply dd-java-agent/instrumentation/spring-webmvc-3.1/src/main/java/datadog/trace/instrumentation/springweb/HandlerMappingResourceNameFilter.java:17]: Yeah, that was my thought also...  If there's a better generic interface that takes a single argument I'd be happy to use that instead.

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Decorator.java [hunk 2]
```diff
@@ -58,12 +61,32 @@ protected Integer status(final HttpServletResponse httpServletResponse) {
...
+  private void onContext(
+      final AgentSpan span, final HttpServletRequest request, final ServletContext context) {
+    final Object attribute = context.getAttribute("dd.dispatcher-filter");
+    if (attribute instanceof Filter) {
+      final Filter filter = (Filter) attribute;
+      try {
+        request.setAttribute(DD_SPAN_ATTRIBUTE, span);
+        filter.doFilter(request, null, null);
+      } catch (final IOException | ServletException e) {
+      }
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Decorator.java:86]: Do we want log here?  Add a health metric?

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/RunnableInstrumentation.java [hunk 3]
```diff
@@ -37,9 +37,7 @@ public RunnableInstrumentation() {
 
   @Override
   public Map<String, String> contextStore() {
-    final Map<String, String> map = new HashMap<>();
-    map.put(Runnable.class.getName(), State.class.getName());
-    return Collections.unmodifiableMap(map);
+    return singletonMap(Runnable.class.getName(), State.class.getName());
```
Comment (dougqh) [inline dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/RunnableInstrumentation.java:40]: This seems completely unrelated.  While it is small, generally, I'd prefer to keep clean-up to that is relevant to the change at hand.

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Decorator.java [hunk 4]
```diff
@@ -58,12 +61,32 @@ protected Integer status(final HttpServletResponse httpServletResponse) {
...
+   * This was easier and less "hacky" than other ways to add the filter to the front of the filter
+   * chain.
+   */
+  private void onContext(
+      final AgentSpan span, final HttpServletRequest request, final ServletContext context) {
+    final Object attribute = context.getAttribute("dd.dispatcher-filter");
+    if (attribute instanceof Filter) {
+      final Filter filter = (Filter) attribute;
+      try {
+        request.setAttribute(DD_SPAN_ATTRIBUTE, span);
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Decorator.java:83]: Should setAttribute be inside the try?

File: dd-java-agent/instrumentation/spring-webmvc-3.1/src/main/java/datadog/trace/instrumentation/springweb/DispatcherServletInstrumentation.java [hunk 5]
```diff
@@ -79,11 +132,6 @@ public static void stopSpan(
       DECORATE_RENDER.beforeFinish(scope);
       scope.close();
```
Comment (dougqh) [inline dd-java-agent/instrumentation/spring-webmvc-3.1/src/main/java/datadog/trace/instrumentation/springweb/DispatcherServletInstrumentation.java:133]: I realize this didn't change in this PR, but I'll ask anyway.

I'd expect scope.close to be inside of a finally block, but usually, it is not.
I'd actually expect XDecorator.beforeFinish to be inside a finally block as well.

Maybe, the first two lines don't typically raise exceptions; however, we might not know because of the suppress=Throwable.

Basically, I'm concerned that this code is not obviously correct.  From looking around, this seems to be a general problem with our resource handling.

Comment (tylerbenson) [reply dd-java-agent/instrumentation/spring-webmvc-3.1/src/main/java/datadog/trace/instrumentation/springweb/DispatcherServletInstrumentation.java:133]: It's true that exceptions thrown from our decorators could cause problems in many places... That should probably be addressed as a separate issue.

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Decorator.java (hunk 2 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Decorator.java:88]: I added a log statement, even though the exception is never thrown.  If we had a better shared interface we could avoid the try/catch (`filter` doesn't actually need to implement the `Filter` interface).
